### ISSUE TITLE
BUG: result with unknown type oid

### ISSUE THREAD

**Issue by lortabac** [valderman/selda#71]
I have this error message whenever I use a type that is not supported by `selda-postgresql`.

I first had it when I declared a column as `INTEGER` (oid 23). In that case I solved the issue by changing the column type to `BIGINT` (oid 20). Related to #51.

Now I am doing an aggregate query on a `BIGINT` column, and somehow the resulting type has oid 1700, which is not supported by `selda-postgresql`. So I have to use a patched version for the moment.

Either I am doing something wrong, or else I do not understand how I am supposed to work with so few types. I can do a PR to add more types if needed. But I want to be sure that the problem is not somewhere else.

**Comment by phaazon** [comment]
What do you mean it’s wrong for PosgreSQL? All aggregates have type `numeric`. Looking at the doc of `numeric`, it seems like it’s an arbitrary-precision-ish, fixed point type. Its representation in Haskell would be `Ratio` or `Fixed` – `Fixed` seems more natural, I guess?

Maybe @lortabac could show the patch he wrote? I’m not a huge fan of overflowing to be honest.

**Comment by valderman** [comment]
@phaazon Aggregation having the type `Num a => Col a -> Aggr a` in Selda is wrong with respect to PostgreSQL, for the very reason you're referring to. The problem with using `Fixed` is that the concept of fixed point types doesn't exist at all in SQLite, and approximating them using floating point types, even if just for SQLite, would just be a disaster waiting to happen.

@lortabac The issue of forcing the types still remains with `(Num a, Num b) => Col a -> Aggr b`, but now the user gets to decide what type to force. It shouldn't be hard in practice, we just make an explicit cast part of the sum operation. I'm leaning more and more towards this, especially since SQLite throws an overflow error instead of just overflowing quietly which throws the first alternative out the window.

**Comment by valderman** [comment]
Even if we don't want to expose `Fixed` as part of the standard API, we should definitely add it as a postgres result type. People adding new types are definitely going to want it, and having the support in `selda-postgresql` won't imply that other backends support it.

**Comment by lortabac** [comment]
> Even if we don't want to expose Fixed as part of the standard API, we should definitely add it as a postgres result type.

+1

> The issue of forcing the types still remains with (Num a, Num b) => Col a -> Aggr b, but now the user gets to decide what type to force. It shouldn't be hard in practice, we just make an explicit cast part of the sum operation.

Do you mean that the new signature will be

```
sum_ :: (SqlType a, SqlType b, Num a, Num b) => Col s a -> (a -> b) -> Aggr s b
```
?

**Comment by valderman** [comment]
@lortabac No, more like `(Num a, Num b, SqlType a, SqlType b) => Col a -> Aggr b`. The decides which type to force through how the result is used (or explicit type annotations). The explicit cast would only appear in the generated SQL.

**Comment by valderman** [comment]
I'd be super happy if someone wants to tackle this, since I won't have the time to fix it myself for at least the next week . It should be a relatively simple change (just change the types and add the appropriate casts) that doesn't even require mucking around in the backend.

**Comment by lortabac** [comment]
I can do it. If I understood correctly it shouldn't be too difficult.
